db/flush_scheduler.cc

```diff
@@ -34,36 +34,37 @@ void FlushScheduler::ScheduleFlush(ColumnFamilyData* cfd) {
 }
 
 ColumnFamilyData* FlushScheduler::TakeNextColumnFamily() {
-  while (true) {
-    if (head_.load(std::memory_order_relaxed) == nullptr) {
+  Node* node;
+  do {
+    if ((node = head_.load(std::memory_order_acquire)) == nullptr) {
       return nullptr;
     }
+  // dequeue the head
+  } while (!head_.compare_exchange_strong(
```

Comment:
This implementation is susceptible to the ABA problem.

Although, the original implementation suggests that there's some other synchronization mechanism at play here, which I don't have full context on; perhaps it's even OK to use a lock instead.

Reply:
Yes you are right, the concurrency situation here is that push (ScheduleFlush) will be called by different threads concurrently, pop (TakeNextColumnFamily) will called sequentially because it's under protection of DB mutex. In other words, there will be N pushes and 1 pop at the same time. Given that, do you think this implementation is thread safe?

Reply:
One ABA problem I can see is: there are 3 threads T1, T2 and T3:
1. T1 call `ScheduleFlush` and go to sleep after creating new node (line 25)
2. T2 call `TakeNextColumnFamily` and delete the current `head_`.
3. T3 call `ScheduleFlush` to insert a new head, which happen to reuse the same address as the deleted `head_`.
4. T1 resume. The CAS operation success by mistake, since the new and old head node has same address.

Reply:
I'm wondering what's the performance impact if `FlushSchedule` use mutex instead of lock-free stack. Flush operation should be infrequent enough?

Reply:
Thanks @yiwu-arbug, @shixiao for input. After discussing with @siying , we feel it's probably best to keep `FlushScheduler` the way it is right now given that PR https://github.com/facebook/rocksdb/pull/5716 eliminated the race condition in pipelined writes. As for `TrimHistoryScheduler` in PR https://github.com/facebook/rocksdb/pull/5022, since it's not on critical path, we can use a slower implementation which uses mutex to simplify the logic.

Reply:
In that case, shall we leave comment in `FlushScheduler` stating its thread unsafeness?

Reply:
It's already mentioned in the .h file: https://github.com/facebook/rocksdb/blob/master/db/flush_scheduler.h#L23-L24

Reply:
ok